clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
@hachikuji Thanks for the explanation. It sounds that you are worrying about the consecutive bad records. If that happens, it is true that the position won't be updated but stick to the first bad message until all the bad records are skipped. In this case updating position to anything other than the first bad position would be inappropriate because we do not really know the offsets of the bad records.

If I understand correctly, we are on the same page that this patch solves the problem it wants to address, i.e. ensure the records are delivered before the bad records. So our discussion is all about the follow up patch on not skipping the bad records.

On that I completely agree that we should just stick to that bad record and ask users to explicitly seek over it. But as you noticed, it is a behavioral change and probably worth discussing it in the community before we make that change (BTW, we do see some applications that just want to ignore the bad records if there are any. There are probably such use case in the community as well). So as for now, we may want to keep the behavior as is.

This would not be a problem if we stick to the bad position, but just from my own curiosity:
> This makes the consumer less functional since the behavior of a fetch at the current offset depends on the history of previous fetches.
Is this true? The exception is only associated with the corresponding bad record. In your example, if an exception is thrown from offset 1, that exception will be cleared. Fetching from offset 2 will only see an exception if offset 2 is a bad record, regardless of whether offset 1 was a bad record or not. So it seems not really depending on the previous exception.